### DISCUSSION TITLE
[Contributing] I want to make the What's New notification openable from the Settings page, but I need some guidance

### DISCUSSION THREAD

**Discussion by stamminator** [DevToys-app/DevToys#407] [Ideas]
_Disclaimer: This is my first time looking at UWP or even XAML code of any kind (my background is ASP.NET & JS), so any and all critiques are welcome. I'm here to learn and hopefully contribute._

I'd love to be able to open the post-update What's New notification on demand from the settings page so you can view it again after you close it. Currently, this functionality lives in `MainPageViewModel`, and the strings passed into `NotificationService` live in `MainPage.resw`. Here are the approaches I've thought of so far:

1. Duplicate the needed strings into `Settings.resw`
   - Bad idea: DRY
2. Access the needed `MainPage.resw` strings from within `SettingsToolViewModel`
   - This seems like poor separation of concerns and a violation of DevToys's modularity, _unless_ MainPage is to be seen as a kind of top-level page that other tools can access, but this doesn't seem to be the intended design.
3. Move the strings to a shared place like `Common.resw`
   - Better, but the strings currently in `Common.resw` seem to indicate that this kind of shared functionality between two pages isn't its intended use case, but is rather for lower level component reusage. Maybe I'm wrong though.

These first three approaches would all share the need to duplicate the following bit of code (more or less) between `MainPageViewModel` and `SettingsToolViewModel`, but maybe it's so minimal that it's not a big deal. Thoughts?
```cs
PackageVersion v = Package.Current.Id.Version;
string? currentVersion = $"{v.Major}.{v.Minor}.{v.Build}.{v.Revision}";

_notificationService.ShowInAppNotification
(
    Strings.GetFormattedNotificationReleaseNoteTitle(currentVersion),
    Strings.NotificationReleaseNoteActionableActionText,
    () =>
    {
        ThreadHelper.ThrowIfNotOnUIThread();
        Windows.System.Launcher.LaunchUriAsync(new Uri("https://github.com/veler/DevToys/releases")).AsTask().Forget();
    },
    await AssetsHelper.GetReleaseNoteAsync()
);
```
And for the last approach...

4. Create a service that's basically the above code, and this service is passed into both `MainPageViewModel` and `SettingsToolViewModel` and consumed.
   - Same original issue: where should the resource strings go? A new `.resw` file and all its localizations just for this service? Seems a little bloated for just one method that's only used in two places.
   - This service would itself depend on `NotificationService`, and I frankly have no clue how the dependency injection for these services works.

Which approach should I use? Is there something I'm not considering? Any advice is appreciated.

**Comment by stamminator** [comment]
@veler Could I bother you for some input on this?

**Comment by veler** [comment]
Hello @stamminator ,

Sorry for the late answer. :) I completely missed your initial message.

There's something I don't think you mentioned in your initial message: what the UI should be? From what I understand from your explanation, it sounds like you'd like to show the notification itself, instead of a modal dialog (for example), which is fine I think. :)

To do this, I can think of a 5th approach which:
1. Doesn't need to duplicate some code.
2. Solves the issue with the strings.

The idea is to make `SettingsToolViewModel` sending a message to `MainPageViewModel` asking it to show the notification.
For this, you can use the `Messenger`. Some documentation here: https://github.com/CommunityToolkit/MVVM-Samples/blob/master/docs/mvvm/Messenger.md

I'd like to invite you to look at the class `NavigateToMarkdownPreviewHtmlMessage`.
https://github.com/veler/DevToys/blob/0465655465fcb842153184e1acc09a07b6d6b2af/src/dev/impl/DevToys/Messages/NavigateToMarkdownPreviewHtmlMessage.cs#L5-L13

This class is a "message" that we use for communicating between the View Model and a View. A message can also be used for communicating between View Models ;-) .
Also, a message can simply be a class without any field or property. Just a strong type. :) 
In order to stay educational and let you discover a bit by yourself, I invite you to look at every place where this class is referenced in the solution. You will be able to find where we "send" the message and where the message is being "received". You can follow this pattern to make the Settings view model sending a defined message to the Main view model, and make the Main view model showing the notification at this moment.

Let me know if you have any question. :)

**Reply by stamminator** [reply]
Thanks for the detailed response! I will dig around and study the existing usages of these messages.
